render-test/manifest_parser.cpp

```diff
@@ -0,0 +1,403 @@
+                                                         const std::regex& regex,
+                                                         bool glyphsPath) const {
+    mbgl::filesystem::path file = std::regex_replace(url, regex, assetPath + parent);
+    if (mbgl::filesystem::exists(file.parent_path())) {
+        return removeURLArguments(file.string());
+    }
+
+    if (glyphsPath && mbgl::filesystem::exists(file.parent_path().parent_path())) {
+        return removeURLArguments(file.string());
+    }
+
+    return {};
+}
+
+namespace {
+std::vector<std::pair<std::string, std::string>> parseIgnores(const std::vector<mbgl::filesystem::path>& ignoresPaths) {
+    std::vector<std::pair<std::string, std::string>> ignores;
+    for (const auto& path : ignoresPaths) {
+        auto maybeIgnores = readJson(path);
+        if (!maybeIgnores.is<mbgl::JSDocument>()) {
+            continue;
+        }
+        for (const auto& property : maybeIgnores.get<mbgl::JSDocument>().GetObject()) {
+            const std::string ignore = {property.name.GetString(), property.name.GetStringLength()};
+            const std::string reason = {property.value.GetString(), property.value.GetStringLength()};
+            ignores.emplace_back(std::make_pair(ignore, reason));
+        }
+    }
+
+    return ignores;
+}
+
+std::vector<mbgl::filesystem::path> getTestExpectations(mbgl::filesystem::path testPath,
+                                                        const mbgl::filesystem::path& testsRootPath,
+                                                        std::vector<mbgl::filesystem::path> expectationsPaths) {
+    std::vector<mbgl::filesystem::path> expectations{std::move(testPath.remove_filename())};
+    const auto& defaultTestExpectationsPath = expectations.front().string();
+
+    const std::regex regex{testsRootPath.string()};
+    for (const auto& path : expectationsPaths) {
+        expectations.emplace_back(std::regex_replace(defaultTestExpectationsPath, regex, path.string()));
+        assert(!expectations.back().empty());
+    }
+
+    return expectations;
+}
+
+mbgl::filesystem::path getValidPath(const std::string& rootPath, const std::string& path) {
+    const static mbgl::filesystem::path BasePath{rootPath};
+    mbgl::filesystem::path result{path};
+    if (result.is_relative()) {
+        result = BasePath / result;
+    }
+    if (mbgl::filesystem::exists(result)) {
+        return result;
+    }
+    mbgl::Log::Warning(mbgl::Event::General, "Invalid path is provoided inside the manifest file: %s", path.c_str());
+    return mbgl::filesystem::path{};
+}
+
+} // namespace
+
+mbgl::optional<Manifest> ManifestParser::parseManifest(const std::string& rootPath,
+                                                       const std::vector<std::string>& testNames,
+                                                       const std::string& testFilter) {
+    static Manifest manifest;
+    auto filePath = mbgl::filesystem::path(rootPath);
```

Comment:
Let's pass the `manifestPath` from the cmd line

Reply:
and it looks like we do not need `rootPath` in that case

Reply:
`rootPath` is necessary in case the paths inside manifest are relative.